Implementing namespacing (#524)
unicode-org/message-format-wg: spec/formatting.md

1. If the _expression_ includes an _operand_, resolve its value.
   If this fails, use a _fallback value_ for the _expression_.
2. Resolve the _identifier_ of the _function_ and, based on the starting sigil,
   find the appropriate function implementation to call.
   If the implementation cannot find the function, emit an Unknown Function error
   and use a _fallback value_ for the _expression_.

>    An implementation MAY emit an Unknown Function error if the implementation    
>    does not support the _namespace_ of the _identifier_
>    or the _namespace_ does not support the named _function_.

Author: You're correct. This should be `MUST`. I think the idea was that implementations are not required to support installable namespaces, but in that case it should say something like:

```suggestion
An implementation MUST emit an Unknown Function error
if the implementation does not support _namespaces_
or if the _namespace_ of the _identifier_ is not recognized by the implementation
or the _namespace_ does not support the named _function_.
```

   Implementations are not required to implement _namespaces_ or installable
   _function registries_.

4. If the _expression_ includes _options_, resolve the _options_ to a mapping
   of string identifiers to values.
   For each _option_:
   - Resolve the _identifier_ of the _option_.
   - If the _option_'s _identifier_ already exists in the resolved mapping of _options_,
     emit a Duplicate Option Name error.
   - If the _option_'s right-hand side successfully resolves to a value,
     bind the _identifier_ of the _option_ to the resolved value in the mapping.
   - Otherwise, do not bind the _identifier_ of the _option_ to any value in the mapping.

> 4. Call the function implementation with the following arguments:

Author: Good callout

> 4. Call the function implementation with the following arguments:

Reviewer: The appropriate-error emission is already taken care of during the resolution of the right-hand value, in Variable Resolution. So if it's noted here, we should not say something like "emit an error" here because that would mean that it needs to be emitted twice.

> 4. Call the function implementation with the following arguments:

Reviewer: @aphillips Our messages crossed paths; see my comment above for a problem this language introduces.

> 4. Call the function implementation with the following arguments:

Author: See change below.

   If this fails, use a _fallback value_ for the _expression_.
2. Resolve the _identifier_ of the _function_ and, based on the starting sigil,
   find the appropriate function implementation to call.
   If the implementation cannot find the function,
   or if the _identifier_ includes a _namespace_ that the implementation does not support,
   emit an Unknown Function error
   and use a _fallback value_ for the _expression_.

   Implementations are not required to implement _namespaces_ or installable
   _function registries_.

3. Resolve the _options_ to a mapping of string identifiers to values.
   If _options_ is missing, the mapping will be empty.
   For each _option_:
   - Resolve the _identifier_ of the _option_.
   - If the _option_'s _identifier_ already exists in the resolved mapping of _options_,
     emit a Duplicate Option Name error.
   - If the _option_'s right-hand side successfully resolves to a value,
     bind the _identifier_ of the _option_ to the resolved value in the mapping.
   - Otherwise, bind the _identifier_ of the _option_ to an unresolved value in the mapping.
     (Note that an Unresolved Variable error will have been emitted.)

> 4. Remove from the resolved mapping of _options_ any binding for which the value is an unresolved value.

Reviewer: This is a small fix unrelated to namespacing that's bundled in here to ensure that if
```
{:foo opt=$none opt=bar}
```

is formatted with `$none` not resolving, not only will that cause an Unresolved Variable error to be emitted, but that a Duplicate Option Name error will also be emitted -- the latter was previously skipped.

The behaviour around options with errors is unchanged from the previous; it's just expressed a bit differently. If you think that ought to be changed, then that should go in a different PR as this is scope-creepy enough already.

> 4. Remove from the resolved mapping of _options_ any binding for which the value is an unresolved value.

Reviewer 2: I filed #545 to continue this.

1. If the _expression_ includes an _operand_, resolve its value.
   If this fails, use a _fallback value_ for the _expression_.
2. Resolve the _identifier_ of the _function_ and, based on the starting sigil,
   find the appropriate function implementation to call.

>    If the implementation cannot find the function, emit an Unknown Function error

Reviewer: @aphillips Would this change (together with removing the next paragraph) be minimally sufficient to covering all cases? Not supporting _any_ namespaces and not supporting _this_ namespace are pretty much the same at this point.

```suggestion
   If the implementation cannot find the function,
   or if the _identifier_ includes a _namespace_ that the implementation does not support,
   emit an Unknown Function error
```

   and use a _fallback value_ for the _expression_.

>    An implementation MUST emit an Unknown Function error
>    if the implementation does not support _namespaces_
>    or if the _namespace_ of the _identifier_ is not recognized by the implementation
>    or the _namespace_ does not support the named _function_.
> 

Reviewer: I still don't see what this is adding to the preceding "If the implementation cannot find the function, emit an Unknown Function error". What sort of a situation would lead to that sentence not being sufficient to also cover everything added here?

Also, the unconditionality of its start now says that an implementation MUST "implement _namespaces_", or always emit an error at this point.

```suggestion
```

Author: it should say "if a namespace is present and the implementation does not support namespaces". Is that a sufficiently separate consideration from "can't find the function"?

Reviewer: The question I'm trying to ask is, if a namespace is present and the implementation does not support namespaces, how can it find the corresponding function?

Author: It can't, which is they the MUSTard says "emit an error if you see a namespace and don't support namespaces"

Reviewer: Now that we have 023b5a8, could this be removed?

1. If the _expression_ includes an _operand_, resolve its value.
   If this fails, use a _fallback value_ for the _expression_.
2. _Resolve the name_ of the _function_ and, based on the starting sigil,
   find the appropriate function implementation from the _function registry_.
   If the registry does not define an implementation for this _name_,
   emit an Unknown Function error
   and use a _fallback value_ for the _expression_.
3. If the _expression_ includes _options_, resolve the _options_ to a mapping
   of string identifiers to values.
   For each _option_:
   - _Resolve the name_ of the _option_.
   - If the _option_'s _name-part_ already exists in the resolved mapping of _options_,
     emit a Duplicate Option Name error.
   - If the _option_'s right-hand side successfully resolves to a value,

>      bind the _name-part_ of the _option_'s to the resolved value in the mapping.

Reviewer: Is this saying that formatting an expression `{:func one:foo=42 two:foo=13}` would generate an error? That would be very surprising, as we'd effectively be saying that the `ns:` part of an option name is completely meaningless.

Author: See: the description of this PR 😀. It does currently mean exactly this. 

I don't think it makes it meaningless, but much depends on how namespaced things (currently only functions and options) are implemented. If the `ns:` says, effectively, where to find the registry entry for validation but not, for example, the invocation, then it probably *is* an error to have `one:foo`/`two:foo` (you can only call one function and that one either expects the `foo` from `one:` or the `foo` from `two:`, but not both. This also alleviates the function from having to figure out that `one:foo` means "my version of `foo`". If the namespace is programmable, the called function has no way of knowing what it is.

Is there a function that would expect two separately-namespaced options of the otherwise-same name?

Reviewer 2: > If the namespace is programmable, the called function has no way of knowing what it is.

I think that *function* namespaces and *option* namespaces should be controlled by different actors, because they serve different purposes.

* Function namespaces are defined **by i18n engineers** who integrate an MF2 implementation into a codebase. They help disambiguate between different packages providing formatters of the same name.
* Option namespaces are defined **by implementors of add-ons** to already existing functions. For example, `:datetime ns:skeleton=YYY` means that the implementation of `datetime` conforms to the default registry, and on top of that, it offers an extra option, `ns:skeleton`.

Note that functions themselves are a kind of a namespace for their options: In both `ns:func foo=bar`, `foo` isn't spelled as `ns:foo` because it's already effectively namespaced by the fact of being an option to `ns:func`. It's the add-on options that need namespacing, and that namespacing should not be programmable.

> Is there a function that would expect two separately-namespaced options of the otherwise-same name? (and, yes, I know this is different from the normal behavior of "namespaces")

As you note, two separately namespaced options are unlikely, because ultimately, there's only one implementation being invoked anyways.

However, we should consider how namespacing impacts other *unnamespaced* options. The namespacing of add-on options guarantees forward compatibility, in case the function being extended adds a new option in the future.

This is why functions should be able to see the namespaces of all their options.

Reviewer 2: In other words: we probably don't need `one:foo=value` and `two:foo=value` to be different, but we do need `foo=value` and `one:foo=value` to be different because o forward compatibility of the add-on mechanism.

The easiest way to achieve it is to pass options together with their namespaces to the function during invocation. This in turn makes `one:foo=value` and `two:foo=value` different options.

Reviewer: I agree with @stasm. I would expect that the `:func` implementation to get called with an options bag
```js
{ 'one:foo': '42', 'two:foo': '13' }
```

Author: Done, although I added text permitting an implementation to do processing of the options.

1. If the _expression_ includes an _operand_, resolve its value.
   If this fails, use a _fallback value_ for the _expression_.
2. _Resolve the name_ of the _function_ and, based on the starting sigil,
   find the appropriate function implementation from the _function registry_.
   If the implementation does not support the _namespace_ of the _name_

>    or the _namespace_ does not support the _name-part_ named _function_
>    or the registry does not otherwise support the _name_,

Reviewer: ```suggestion
   or the _namespace_ does not support the _name-part_ named _function_,
```

Referring to the registry here would require it to concretely exist during the formatting.

Author: This is the part that I *didn't* change from the original.

Reviewer: Lol, you're right. So let's fix an earlier mistake.

1. If the _expression_ includes an _operand_, resolve its value.
   If this fails, use a _fallback value_ for the _expression_.
2. _Resolve the name_ of the _function_ and, based on the starting sigil,
   find the appropriate function implementation to call.
   If the implementation cannot find the function, emit an Unknown Function error
   and use a _fallback value_ for the _expression_.

> 
>    An implementation MAY emit an Unknown Function error if the implementation    
>    does not support the _namespace_ of the _name_
>    or the _namespace_ does not support the _name-part_ named _function_.

Reviewer: This sentence does not add anything to the previous one. If either of the conditions here are true, then it's also true that the implementation "cannot find the function", and will therefore emit the error.

```suggestion
```

> 
>    Implementations are not required to implement _namespaces_ or installable
>    _function registries_.

Reviewer: This seems out of place. If we do want to mention something like this, it should not be here, in the middle of an algorithmic sequence of steps.

```suggestion
```

> 4. If the _expression_ includes _options_, resolve the _options_ to a mapping
>    of string identifiers to values.

Reviewer: Including the conditional here means that if there are no options, no (empty) mapping is defined. We should drop the conditional and rely on the implicit skipping of the "for each" loop as previously.

Also, the numbering got confused again.

```suggestion
3. Resolve the _options_ to a mapping of string identifiers to values.
```

   For each _option_:
   - _Resolve the name_ of the _option_.
   - If the _option_'s _name_ already exists in the resolved mapping of _options_,
     emit a Duplicate Option Name error.
   - If the _option_'s right-hand side successfully resolves to a value,
     bind the _name_ of the _option_ to the resolved value in the mapping.
   - Otherwise, do not bind the _name_ of the _option_ to any value in the mapping.

> 5. Call the function implementation with the following arguments:

Reviewer: ```suggestion
4. Call the function implementation with the following arguments:

```

   An implementation MAY pass additional arguments to the function,
   as long as reasonable precautions are taken to keep the function interface
   simple and minimal, and avoid introducing potential security vulnerabilities.
   Implementations SHOULD use an implementation-defined _namesapce_ for

>    any additional arguments exposed to users as _options_.

Author: Yeah, this needed more work. Have another look.

>    An implementations MAY define its own functions or allow custom functions
>    to be defined by users.

Author: No. They should be separate statements.

```suggestion
An implementation MAY define its own functions.
An implementation MAY allow custom functions to be defined by users.
```

   Function access to the _formatting context_ SHOULD be minimal and read-only,

>    and execution time SHOULD be limited.

Author: I didn't really change the requirement, but agree that we should.

   Implementation-defined _functions_ SHOULD use an implementation-defined _namespace_.
   User-defined _functions_ SHOULD be permitted to define the _namespace_ to use

>    in name resolution.

Reviewer: I still don't understand what this means.

8. If the call succeeds,
   resolve the value of the _expression_ as the result of that function call.
   If the call fails or does not return a valid value,
   emit a Resolution error and use a _fallback value_ for the _expression_.

### Name Resolution

<dfn title="resolve the name">**_Name resolution_**</dfn> is the determination of
the _name_ and the _namespace_ of the name for a given _function_
or _option_.

> To determine the _name_ of an item, the following steps are taken:

Author: The colon is not part of the namespace or the name-part. It's just a separator.

   as long as reasonable precautions are taken to keep the function interface
   simple and minimal, and avoid introducing potential security vulnerabilities.

   Implementations MAY expose _options_ in an implementation-defined _namespace_

>    in addition to or superceding those found in the default registry.

Reviewer 3: ```suggestion
   in addition to or superseding those found in the default registry.
```

   An implementation MAY define its own functions.
   An implementation MAY allow custom functions to be defined by users.

   Function access to the _formatting context_ MUST be minimal and read-only,
   and execution time SHOULD be limited.
   Implementation-defined _functions_ SHOULD use an implementation-defined _namespace_.
   User-defined _functions_ SHOULD be permitted to define the _namespace_ to use

>    in name resolution.

Reviewer 3: Does this mean that a user can define their own namespaces (e.g. introduce new namespaces with arbitrary names), or that they can choose a namespace from a set of existing namespaces?

5. If the call succeeds,
   resolve the value of the _expression_ as the result of that function call.
   If the call fails or does not return a valid value,
   emit a Resolution error and use a _fallback value_ for the _expression_.

### Name Resolution

<dfn title="resolve the name">**_Name resolution_**</dfn> is the determination of
the _name-part_ and, where present, the _namespace_ of the _name_ for a given _function_
or _option_.

To resolve the _name_ of an item, the following steps are taken:

1. Let the _name_ be the string to be processed, less any preceeding
   sigils or markers.
   For example, in the _expression_ `{:function}`, the string `function`
   is the _name_ string.
2. Determine if _name_ contains the character U+003A COLON `:`:
   - If _name_ contains a colon, let _namespace_ be the
      substring of _name_ up to (but not including)

>       the colon and _name-part_ be the substring of _name_ following the colon;

Author: Removed the ability of colon to appear in `name`. Note that this makes our names match NCName (which I cite in this revision). This helps reduce the "sorta like an" XML name thing.

   - Else let _namespace_ be an empty value (null, void, or empty string)
     and _name-part_ be the _name_.
3. If the _name-part_ is empty, return a Syntax error.
4. Return _namespace_ and _name-part_.

Implementations are not required to support the installation or resolution
of different namespaces.
How namespaces are defined, provisioned, or handled is also implementation defined.
It is important to note that the _namespace_ is not considered part of the _name_
after resolution. 
The _namespace_ MAY affect which function is called.

Example:
> Suppose the ICU implementation of MessageFormat added a custom option `skeleton`
> to the `:datetime` function.
> Users of that implementation could invoke the datetime formatter with an

> > expression similar to:

Reviewer 3: ```suggestion
> pattern similar to:
```
(What follows is a pattern; alternately, you could leave this as-is and change the example to `{$date :datetime icu:skeleton=yMMMd}`)

> ```
> {{Today is {$date :datetime icu:skeleton=yMMMd}.}}
> ```

> > The `name` of the `datetime` function is resolved as:
> > > `namespace`: (blank)
> > > `name`: `datetime`
> > The `name` of the option in the expression is resolved as:
> > > `namespace`: `icu`
> > > `name`: `skeleton`

Reviewer 4: There's some Markdown mangling here: https://github.com/unicode-org/message-format-wg/blob/1dc97bfcd95494e80de5a2b1e5ee62681f4c92e4/spec/formatting.md#name-resolution
```suggestion
> The `name` of the `datetime` function is resolved as:
> * `namespace`: (blank)
> * `name`: `datetime`
> 
> The `name` of the option in the expression is resolved as:
> * `namespace`: `icu`
> * `name`: `skeleton`
```

> > > `name`: `skelelton`

Reviewer 3: ```suggestion
> > `name`: `skeleton`
```

   as long as reasonable precautions are taken to keep the function interface
   simple and minimal, and avoid introducing potential security vulnerabilities.

   Implementations MAY expose _options_ in an implementation-defined _namespace_
   in addition to or superseding those found in the default registry.

   An implementation MAY define its own functions.
   An implementation MAY allow custom functions to be defined by users.

   Function access to the _formatting context_ MUST be minimal and read-only,
   and execution time SHOULD be limited.

>    Implementation-defined _functions_ SHOULD use an implementation-defined _namespace_.

Author: I'm thinking of going back to the design document and adding some use cases to illustrate these. In this case, I mean "non-default registry functions". I will add text to clarify.

---

Suppose you're Mihai and you're creating the ICU4J implementation. The `:datetime` function in the default registry might be backed by `com.ibm.icu.text.DateFormat` as an implementing class. There's no need to prefix `:datetime` for that. That's just implementation.

Mihai might want to expose `skeleton` though. That's not (currently) a default registry option. So he'd prefix it as `icu:skeleton` in his implementation.

Suppose he then goes on to implement `com.ibm.icu.text.DateIntervalFormat` as a function. That's not in the default registry, so it gets `:icu:dateinterval` and its options might or might not be namespaced.

Suppose Mihai goes on to implement an SPI for custom functions. Suppose I write my own date formatter. It doesn't replace the one in `:dateformat`, so it gets a prefix like `:app:dateformat`. The standard options might not be prefixed, but my custom options would be. ICU's options would probably be unrecognized to me. So you could see expressions like:

```
{$now :datetime icu:skeleton=yMMMd}
{$now :icu:dateinterval end=$then skeleton=yMMM}
{$now :app:datetime dateStyle=short :app:wonderfulness=high}
```

Suppose you're a tool implementer working with messages written for Mihai's implementation with my add-on. When you see a namespace prefix, that tells you that the function or the option is non-standard. The prefix might tell you where to look for the add-on registry or at least be something that you can key locally. Tool consumption is a key reason for namespacing.

>    Implementation-defined _functions_ SHOULD use an implementation-defined _namespace_.

Reviewer 2: Adding these use-cases in form of such user journeys sounds great. I'll be happy to contribute. Open a new PR?

>    Implementation-defined _functions_ SHOULD use an implementation-defined _namespace_.

Reviewer: Might be best to add a subsection to "Function Resolution" that collects these directions about what's possible and recommended for implementations and users to do with functions?

```suggestion
```

   as long as reasonable precautions are taken to keep the function interface
   simple and minimal, and avoid introducing potential security vulnerabilities.

   Implementations MAY expose _options_ in an implementation-defined _namespace_

>    in addition to or superseding those found in the default registry.

Reviewer: I don't understand how including or not including this sentence changes what implementations and users are able or allowed to do.

Author: Rmoeved

   as long as reasonable precautions are taken to keep the function interface
   simple and minimal, and avoid introducing potential security vulnerabilities.

>    An implementation MAY define its own functions.
>    An implementation MAY allow custom functions to be defined by users.
> 

Reviewer: This too is out of place, and should be stated elsewhere.

```suggestion
```

Author: Where?

>    An implementation MAY perform additional processing before calling the function.
>    For example, it can modify the _name_ of an _option_ to match the function's
>    input expectations (such as by removing the _namespace_ prefix from the _name-part_).
> 

Reviewer: This feels a bit weird, and I don't think it should be allowed. The functions that an implementation will end up calling should be considered rather specific to that implementation, and therefore fully capable of dealing with their options without preprocessing.

For example, let's say an implementation provides an `icu:number` function that's mostly the same except for the core `number` but it adds a few options. That function may internally support both `icu:skeleton` and `skeleton` as synonyms for each other, but figuring that out ought to be on the function, rather than the implementation. E.g. detecting that both should not be defined at the same time ought to be a function concern.

```suggestion
```

Author: I think of the function and the implementation of the function interchangeably. This permits an implementation to say to plug-in functionality that the namespacing will be managed for the function. But we could leave it out.

   Function access to the _formatting context_ MUST be minimal and read-only,
   and execution time SHOULD be limited.
   
   Implementation-defined _functions_ SHOULD use an implementation-defined _namespace_.

> 7. If the call succeeds,

Reviewer: ```suggestion
5. If the call succeeds,
```
